vmware_host: add reconnect and add_or_reconnect states

tchernomax opened the pull request:
##### SUMMARY

Add "reconnect" and "add_or_reconnect" choices for "state" in vmware_host module.

* reconnect: reconnect an esxi to a vcenter (imply it is present).
* add_or_reconnect: do the same but add the esxi if absent.

Also:
* tag the cluster_name as required (because it is).
* tag esxi_username and esxi_password as not required because
    they aren't when the esxi isn't added.

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

vmware_host

##### ANSIBLE VERSION

```
ansible 2.5.0 (vmware_host 7fe1b45ba6) last updated 2017/09/19 23:52:13 (GMT +200)
  config file = /home/max/.ansible.cfg
  configured module search path = ['/home/max/.ansible/modules']
  ansible python module location = /home/max/Documents/repositories-github/ansible/lib/ansible
  executable location = /home/max/Documents/repositories-github/ansible/bin/ansible
  python version = 3.6.2 (default, Jul 20 2017, 03:52:27) [GCC 7.1.1 20170630]
```

Commit:
vmware_host: add reconnect and add_or_reconnect states

Add "reconnect" and "add_or_reconnect" choices for "state".

* reconnect: reconnect an esxi to a vcenter (imply it is present).
* add_or_reconnect: do the same but add the esxi if absent.

Also:
* tag the cluster_name as required (because it is).
* tag esxi_username and esxi_password as not required because
    they aren't when the esxi isn't added.

Akasurde commented:
@tchernomax Thanks for the PR. Could you please add integration tests for this change ? You can start [here](https://github.com/ansible/ansible/tree/devel/test/integration/targets/vmware_host)

Commit:
vmware_host: add + prepare/document integration tests

vmware_host module

Add integration test for the add part of "add_or_reconnect" state.

Prepare and document integration tests for the reconnect part
of "add_or_reconnect" state and "reconnect" and "absent" states.

Currently we can't test those states as ReconnectHost_Task (for
"reconnect") and EnterMaintenanceMode_Task (for "absent") aren't
implemented yet in vcsim (from vmware/govmomi)

tchernomax commented:
@Akasurde I add one test and documents some others but currently [vcsim](https://github.com/vmware/govmomi/tree/master/vcsim#govcsim---a-vcenter-and-esxi-api-based-simulator) only support the following methods and we need ReconnectHost_Task and EnterMaintenanceMode_Task to fully test the module.
```
max@laptop % curl -k 'https://user:pass@localhost:443/about'
{
  "Methods": [
    "AcquireGenericServiceTicket",
    "AddDVPortgroup_Task",
    "AddHost_Task",
    "AddPortGroup",
    "AddStandaloneHost_Task",
    "AddVirtualSwitch",
    "CancelWaitForUpdates",
    "CloneVM_Task",
    "CreateClusterEx",
    "CreateContainerView",
    "CreateDVS_Task",
    "CreateDatacenter",
    "CreateFilter",
    "CreateFolder",
    "CreateLocalDatastore",
    "CreateNasDatastore",
    "CreatePropertyCollector",
    "CreateResourcePool",
    "CreateStoragePod",
    "CreateVApp",
    "CreateVM_Task",
    "CurrentTime",
    "DeleteDatastoreFile_Task",
    "DestroyPropertyCollector",
    "DestroyView",
    "Destroy_Task",
    "DisableRuleset",
    "EnableRuleset",
    "FindByDatastorePath",
    "FindByInventoryPath",
    "FindByUuid",
    "FindChild",
    "Login",
    "Logout",
    "MakeDirectory",
    "MoveDatastoreFile_Task",
    "MoveIntoFolder_Task",
    "PowerOffVM_Task",
    "PowerOnVM_Task",
    "QueryAssignedLicenses",
    "QueryOptions",
    "ReconfigVM_Task",
    "ReconfigureDvs_Task",
    "RefreshDatastore",
    "RegisterVM_Task",
    "RelocateVM_Task",
    "RemovePortGroup",
    "RemoveVirtualSwitch",
    "Rename_Task",
    "RetrieveProperties",
    "RetrievePropertiesEx",
    "RetrieveServiceContent",
    "SearchDatastoreSubFolders_Task",
    "SearchDatastore_Task",
    "ShutdownGuest",
    "UnregisterVM",
    "WaitForUpdates",
    "WaitForUpdatesEx"
  ],
  "Types": [
    "ClusterComputeResource",
    "ComputeResource",
    "ContainerView",
    "Datacenter",
    "Datastore",
    "DistributedVirtualPortgroup",
    "FileManager",
    "Folder",
    "HostDatastoreBrowser",
    "HostDatastoreSystem",
    "HostFirewallSystem",
    "HostNetworkSystem",
    "HostSystem",
    "LicenseAssignmentManager",
    "LicenseManager",
    "Network",
    "OptionManager",
    "PropertyCollector",
    "ResourcePool",
    "SearchIndex",
    "ServiceInstance",
    "SessionManager",
    "StoragePod",
    "Task",
    "TaskManager",
    "ViewManager",
    "VirtualMachine",
    "VmwareDistributedVirtualSwitch"
  ]
}
```

mtnbikenc commented:
@ansibot 
@jcpowermac @mtnbikenc would like to be removed from automatic cc on vmware prs, please

Akasurde commented:
@mtnbikenc Should I remove you from VMware maintainers' list ?

Akasurde commented:
@tchernomax Ok.

mtnbikenc commented:
@Akasurde Yes, please remove me from the VMware maintainers list.  I've asked before but seems I keep getting put back on.  Thanks!

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host.py:
```diff
@@ -31,7 +32,7 @@
     description:
     - Name of the datacenter to add the host.
-    required: yes
+    required: no
```
datacenter_name is required parameter.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host.py:
```diff
@@ -43,16 +44,30 @@
+    - absent
+    - add_or_reconnect
+    - reconnect
+    required: no
```
required:false is default so no need to specify.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host.py:
```diff
@@ -199,17 +282,17 @@ def main():
-        esxi_password=dict(type='str', required=True, no_log=True),
-        state=dict(type='str', default='present', choices=['absent', 'present'])
-    )
+        esxi_username=dict(type='str', required=False),
```
esxi_username, esxi_password are required options, so maybe you can use  ```required_if```

tchernomax replied:
I didn't know `required_if`, thanks for the hint. I did the modification.

tchernomax commented:
@jcpowermac @Akasurde I just applied the modifications pointed by @Akasurde and rebased on the current devel branch.

tchernomax commented:
@Akasurde this PR kept the tagged "Changes requested" even if I applied your changes… what should I do ?

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host.py:
```diff
@@ -43,16 +44,27 @@
   esxi_username:
     description:
     - ESXi username.
-    required: yes
```
Add note here saying, "esxi_username is no longer required parameter from version 2.5"

This is to maintain backward compatibility.

tchernomax replied:
done https://github.com/ansible/ansible/pull/30582/files#diff-e058a458f41f76acd5fa09c56008b0f0R51

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host.py:
```diff
@@ -43,16 +44,27 @@
   esxi_password:
     description:
     - ESXi password.
-    required: yes
```
Same here.

tchernomax replied:
done https://github.com/ansible/ansible/pull/30582/files#diff-e058a458f41f76acd5fa09c56008b0f0R58

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host.py:
```diff
@@ -199,17 +279,20 @@ def main():
     argument_spec.update(
         datacenter_name=dict(type='str', required=True),
-        cluster_name=dict(type='str'),
+        cluster_name=dict(type='str', required=True),
```
Modify the documentation to reflect this change. Also, add version in "2.5" this behaviour is changed.

tchernomax replied:
done https://github.com/ansible/ansible/pull/30582/files#diff-e058a458f41f76acd5fa09c56008b0f0R39

tchernomax commented:
@Akasurde all requested changes have been done

Akasurde reviewed (approved):
shipit

tchernomax commented:
@jctanner @pdellaert

jctanner commented:
shipit
